gnocchi/rest/aggregates/measures_grouper.py

```diff
@@ -0,0 +1,305 @@
+# distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+# WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+# License for the specific language governing permissions and limitations
+# under the License.
+
+import fnmatch
+import itertools
+import logging
+
+import numpy
+import pecan
+
+from gnocchi import indexer
+from gnocchi.rest.aggregates import exceptions
+from gnocchi.rest.aggregates import processor
+from gnocchi.rest import api
+from gnocchi import storage
+
+LOG = logging.getLogger(__name__)
+
+
+def get_measures_or_abort(references, operations, start,
+                          stop, granularity, needed_overlap, fill):
+    try:
+        return processor.get_measures(
+            pecan.request.storage,
+            references,
+            operations,
+            start, stop,
+            granularity, needed_overlap, fill)
+    except exceptions.UnAggregableTimeseries as e:
+        api.abort(400, e)
+    # TODO(sileht): We currently got only one metric for these exceptions but
+    # we can improve processor to returns all missing metrics at once, so we
+    # returns a list for the future
+    except storage.MetricDoesNotExist as e:
+        api.abort(404, {"cause": "Unknown metrics",
+                        "detail": [str(e.metric.id)]})
+    except storage.AggregationDoesNotExist as e:
+        api.abort(404, {"cause": "Metrics with unknown aggregation",
+                        "detail": [(str(e.metric.id), e.method)]})
+
+
+def _get_measures_by_name(resources, metric_wildcards, operations,
+                          start, stop, granularity, needed_overlap, fill,
+                          details):
+    references = []
+    for r in resources:
+        references.extend([
+            processor.MetricReference(m, agg, r, wildcard)
+            for wildcard, agg in metric_wildcards
+            for m in r.metrics if fnmatch.fnmatch(m.name, wildcard)
+        ])
+
+    if not references:
+        raise indexer.NoSuchMetric(set((m for (m, a) in metric_wildcards)))
+
+    response = {
+        "measures": get_measures_or_abort(
+            references, operations, start, stop, granularity,
+            needed_overlap, fill)
+    }
+    if details:
+        response["references"] = set((r.resource for r in references))
+    return response
+
+
+class MeasureGroup(object):
+    def __init__(self, group_key, resources):
+        self.resources = self.join_sequential_groups(resources)
+        self.group_key = dict(group_key)
+        self.measures = []
+        self.references = None
+
+    def add_measures(self, aggregated_measures, date_map, start, stop,
+                     details):
+        measures = aggregated_measures['measures']['aggregated']
+        if details:
+            self.references = aggregated_measures['references']
+        for measure in measures:
+            self.add_measure(list(measure), date_map, start, stop)
```

Comment:
hmm... the performance of this is a bit concerning. it seems to be going through each timeseries point by point and trying to slot in to a group (in python)

Reply:
got it. (wasn't aware nova kept same id when flavour changes). i do understand the problem you're trying to solve, i guess my question is what the correct behaviour is to attribute ownership when dividing up an measure.

for instance, if we're capturing cpu.time|storage.bytes which is a gauge metric that (for the most part) only grows. if we split, it with current logic say it might end up being:
- 00:00:00, value 0, meta1
- 10:00:00, value 10, meta1
- 11:00:00, value 1.1; meta1(11:00 - 11:06)
- 11:00:00, value 8,8: meta2(11:06 - 11:54)
- 11:00:00, value 1.1, meta3 (11:54 - 12:00)

which would be an odd value considering it was meta1 for 11hrs and meta2 for 48min but the values are near identical. maybe i'm over-complicating this.

Reply:
Got your point, and yes, your are right, this kind of behavior does not make much sense when using gauge metrics, they are cumulative. 

You should not configure metadata sensitive resources using gauge metrics when you have floating metadata (each possible metadata should be a resource itself to gauge make sense), because any time the metadata changes you should reset the measures values to zero (because it means that this is a new one) and pause the old metadata collecting until the resource changes its metadata again to the old value (togging between metadata values).

For these cases, if you want to use gauge metrics, you should create one metric for any possible metadata value in the resource and measure them separated in your data polling service. 

So if this is the case, you could continue using the normal aggregation API without sending the parameter "user_history" in the request, so the default behavior will me maintained.